src/hotspot/share/runtime/synchronizer.cpp

```diff
@@ -415,14 +410,34 @@ bool ObjectSynchronizer::quick_enter(oop obj, JavaThread* current,
     return false;
   }
 
+  if (useHeavyMonitors()) {
+    return false;  // Slow path
+  }
+
   if (LockingMode == LM_LIGHTWEIGHT) {
-    return LightweightSynchronizer::quick_enter(obj, current, lock);
+    LockStack& lock_stack = current->lock_stack();
+    if (lock_stack.is_full()) {
+      // Always go into runtime if the lock stack is full.
+      return false;
+    }
+
+    if (lock_stack.try_recursive_enter(obj)) {
+      // Recursive lock successful.
+      current->inc_held_monitor_count();
+      return true;
```

Comment:
This needs a CacheSetter

```suggestion
      CacheSetter cache_setter(current, lock);
  
      // Recursive lock successful.
      current->inc_held_monitor_count();
      return true;
```

Reply:
Ok, yes, I missed that the fast lock stack case set the monitor.  I was trying to keep the CacheSetter contained in lightweightSynchronizer.cpp.  I'll do some more refactoring.  Thanks for pointing this out.

Reply:
Oh now I see.  If we do a recursive lock, we need to clear_object_monitor_cache for some reason I haven't figured out yet.

Reply:
Yeah I am not 100% happy with the CacheSetter, it at least needs better documentation. The issue is an uninitialised value will (more than likely) look like a ObjectMonitor* to the cache lookup code. So it will break if this thread observers that it is inflate (by for example calling wait) then it will crash when exiting.

It aims to uphold the simple invariant that after ObjectSynchronize::enter the BasicLock contains either the ObjectMonitor locked on or nullptr.